cpp/src/structs/copying/concatenate.cu

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/concatenate.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/concatenate.cuh>
+#include <cudf/detail/get_value.cuh>
+#include <cudf/structs/structs_column_view.hpp>
+#include <memory>
+
+namespace cudf {
+namespace structs {
+namespace detail {
+
+/**
+ * @copydoc cudf::structs::detail::concatenate
+ *
+ */
+std::unique_ptr<column> concatenate(
+  std::vector<column_view> const& columns,
+  cudaStream_t stream                 = 0,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+{
+  // concatenate children.
+  std::vector<std::unique_ptr<cudf::column>> children;
+  children.reserve(columns.size());
+  auto child_index = thrust::make_counting_iterator(0);
+  std::transform(child_index,
```

Comment:
It feels like there's a reusable piece of functionality that can be distilled here. I'd think it would be a common operation if given a `vector<column_view>` of struct columns to want to create a `vector< vector<column_view> >` where sub-vector `i` is all of the `i` children from each original vector. Example:

```
std::vector<column_view> vector_of_structs { S0{A0, B0, C0}, 
                                             S1{A1, B1, C1},
                                             S2{A2, B2, C2}};

vector< vector<column_view> > struct_children { {A0, A1, A2}, {B0, B1, B2}, {C0, C1, C2} };
```

It's effectively like a transpose, but you're "transposing" the children out of a struct column. The way it could be used here in concatenate is like:
```
auto struct_children = tranpose_struct_children(struct_columns);

std::vector< std::unique_ptr<cudf::column> > concatenated_children;
std::transform( struct_children.begin(), struct_children.end(), back_inserter(concatenated_children.begin()), []( vector<column_view> const& children) { return concatenate(children, mr, stream); };
```

Reply:
This makes sense although I think transpose is a little cryptic.  Maybe "extract_ordered_children", or even just "extract_children".

Reply:
Yeah, transpose definitely wasn't meant to be final naming but rather get the idea across.